### PR TITLE
add support for yaml merging via << key

### PR DESCRIPTION
**Pull request by jreinert**

### CODE REVIEW

File: src/yaml/parser.cr [hunk 1]
```diff
@@ -83,7 +83,11 @@ class YAML::Parser
         key = parse_node
         @pull_parser.read_next
         value = parse_node
-        mapping[key] = value
+        if key == "<<"
+          raise "Expected value to be a Hash" unless value.is_a?(Hash)
```
Comment (jhass) [inline src/yaml/parser.cr]: Can we find a more specific exception here?

File: src/yaml/parser.cr [hunk 2]
```diff
@@ -83,7 +83,11 @@ class YAML::Parser
...
+          mapping.merge!(value)
+        else mapping[key] = value
```
Comment (jhass) [inline src/yaml/parser.cr]: Let's add a newline here, that always looks like an empty `elsif` on first glance to me.

**Comment by jhass** [comment]
Shouldn't this rather support [alias nodes](http://yaml.org/spec/1.2/spec.html#id2786196)?

**Comment by jhass** [comment]
We probably should ensure the behaviour outlined by [psych's tests for this](https://github.com/tenderlove/psych/blob/master/test/psych/test_merge_keys.rb)

**Comment by jreinert** [comment]
@jhass sure, the primary use case is alias nodes. But those are tested separately and if merging of mappings is working, so is merging aliases of mappings (transitively).

**Comment by jreinert** [comment]
the psych tests are a good resource. I will update my tests. (It would probably fail the implicit String test as of now)

**Comment by jreinert** [comment]
as long as this:
https://github.com/tenderlove/psych/blob/master/test/psych/test_merge_keys.rb#L173
isn't clear to me I can't implement merging of multiple mappings. I'm confused why the 10 from `BIG` is overriding the 1 from `SMALL` even though `SMALL` is last in the array.

**Comment by jhass** [comment]
> Keys in mapping nodes earlier in the sequence override keys specified in later mapping nodes.

http://yaml.org/type/merge.html

It simply is backwards. See also https://github.com/ruby/ruby/commit/6e5aa6311a5798145daf0c42ce0a757c333f7155

**Comment by jreinert** [comment]
i see. alright then.

File: src/yaml/parser.cr [hunk 3]
```diff
@@ -93,6 +98,14 @@ class YAML::Parser
     value
   end
 
+  private def merge_mapping!(mapping, key, value)
```
Comment (asterite) [inline src/yaml/parser.cr]: It seems this method is never used? Maybe you refactored it a bit and forgot to remove or finish something?

**Comment by asterite** [comment]
Looks good, except for that unused method :+1: 

Is this ">>" standard in YAML? I mean, does it depend on some schema?

**Comment by jreinert** [comment]
@asterite i removed the unused method. Yes "<<" is standard in YAML. See the official docs http://yaml.org/type/merge.html

**Comment by asterite** [comment]
:+1: For me. If someone else agrees we can merge this. /cc @jhass @ysbaddaden @waj

**Comment by jhass** [comment]
LGTM :+1:

**Comment by ysbaddaden** [comment]
Lets merge then :-)

**Pull request merged**